libqtile/widget/base.py

```diff
@@ -847,43 +872,46 @@ class ThreadPoolText(_TextBox):
 
     def __init__(self, text="N/A", **config):
         super().__init__(text, **config)
-        self.add_defaults(ThreadPoolText.defaults)
+        self.add_defaults(BackgroundPoll.defaults)
 
     def timer_setup(self):
-        def on_done(future):
+        asyncio.create_task(self.do_tick())
+
+    def poll(self):
+        """An optional non-async-based method for polling. Will be run as an
+        async future."""
+
+    async def apoll(self):
+        """An optional async-based method for polling."""
+
+    async def do_tick(self):
+        if type(self).apoll != BackgroundPoll.apoll:
+            result = await self.apoll()
+        elif type(self).poll != BackgroundPoll.poll:
+            future = self.qtile.run_in_executor(self.poll)
+            result = await future
+        else:
+            raise Exception(f"widget {self.name} has neither apoll() nor poll() overridden?")
+        if result is not None:
             try:
-                result = future.result()
+                self.update(result)
             except Exception:
-                result = None
-                logger.exception("poll() raised exceptions, not rescheduling")
-
-            if result is not None:
-                try:
-                    self.update(result)
-
-                    if self.update_interval is not None:
-                        self.timeout_add(self.update_interval, self.timer_setup)
-
-                except Exception:
-                    logger.exception("Failed to reschedule timer for %s.", self.name)
-            else:
-                logger.warning("%s's poll() returned None, not rescheduling", self.name)
-
-        self.future = self.qtile.run_in_executor(self.poll)
-        self.future.add_done_callback(on_done)
+                logger.exception("Failed to reschedule timer for %s.", self.name)
+        else:
+            logger.warning("%s's poll() returned None, not rescheduling", self.name)
+        if result is not None:
+            self.update(result)
 
-    def poll(self):
-        pass
+        if result is not None and self.update_interval is not None:
+            await asyncio.sleep(self.update_interval)
+            asyncio.create_task(self.do_tick())
 
     @expose_command()
     def force_update(self):
         """Immediately poll the widget. Existing timers are unaffected."""
         self.update(self.poll())
```

Comment:
I think we need to update this too.

Reply:
Oh, yes, good call.

Reply:
Should be fixed now, it is a boolean parameter about whether or not to sleep+requeue. Yes, boolean parameters are bad, but doing the method resolution again somewhere else and returning the thing, which is sometimes async and sometimes not made type checking hard :)

Reply:
Heh, ok. Finally fixed now :)